File: server/src/collectors/blitzortungLightning.js
/**
 * Blitzortung — citizen-network real-time lightning strikes.
 * https://www.blitzortung.org/en/live_lightning_maps.php
 *
 * Free WebSocket stream (`wss://ws*.blitzortung.org/`) of detected impacts.
 * Pair with `jma-weather` for severe-storm correlation. We register the
 * portal here; a follow-up WS consumer ingests strikes into a feature
 * stream.
 */

import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';
import { fetchHead } from './_liveHelpers.js';

const SOURCE_ID = 'blitzortung-lightning';
const PROBE_URL = 'https://www.blitzortung.org/en/live_lightning_maps.php';

export default async function collectBlitzortungLightning() {
  const live = await fetchHead(PROBE_URL).catch(() => false);
  return intelEnvelope({
    sourceId: SOURCE_ID,
    items: [{
      uid: intelUid(SOURCE_ID, 'portal'),
      title: 'Blitzortung lightning network',
      summary: 'Citizen-detected real-time lightning strikes (WSS feed)',
      link: PROBE_URL,
      language: 'en',
      published_at: new Date().toISOString(),
      tags: ['lightning', 'weather', 'citizen-science', live ? 'reachable' : 'unreachable'],
      properties: { operator: 'Blitzortung.org', reachable: live },
    }],
    live,
    description: 'Blitzortung — real-time lightning detection network',
  });
}

File: native/collectors/sources/blitzortung_lightning.c
/* collectors/environment/sources/blitzortung_lightning.c
 * Port of server/src/collectors/blitzortungLightning.js — fetchHead
 * reachability probe → ONE intel item (uid blitzortung-lightning|portal ==
 * intelUid(SOURCE_ID,'portal')). OTHER portal-status family (lib/probe.h). */
#include "../../source.h"
#include "../../lib/probe.h"
#include "../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>

#define PROBE_URL "https://www.blitzortung.org/en/live_lightning_maps.php"

static int run(const source_ctx *ctx, intel_sink *sink) {
  int live = probe_head(ctx->http, PROBE_URL);
  char now[32]; probe_iso_now(now, sizeof now);

  /* tags: ['lightning','weather','citizen-science', live?'reachable':'unreachable'] */
  cJSON *tags = cJSON_CreateArray();
  cJSON_AddItemToArray(tags, cJSON_CreateString("lightning"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("weather"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("citizen-science"));
  cJSON_AddItemToArray(tags, cJSON_CreateString(live ? "reachable" : "unreachable"));
  char *tj = cJSON_PrintUnformatted(tags);

  /* properties: { operator, reachable } — EXACT JS key order */
  cJSON *p = cJSON_CreateObject();
  cJSON_AddStringToObject(p, "operator", "Blitzortung.org");
  cJSON_AddBoolToObject(p, "reachable", live);
  char *pj = cJSON_PrintUnformatted(p);

  intel_item it = {0};
  it.remote_key   = "portal";                 /* → uid blitzortung-lightning|portal */
  it.title        = "Blitzortung lightning network";
  it.summary      = "Citizen-detected real-time lightning strikes (WSS feed)";
  it.link         = PROBE_URL;
  it.published_at = now;
  it.tags_json    = tj;
  it.properties_json = pj;
  int rc = sink->emit(sink, &it);

  free(tj); free(pj);
  cJSON_Delete(tags); cJSON_Delete(p);
  fprintf(stderr, "[blitzortung-lightning] probe reachable=%d\n", live);
  return rc >= 0 ? 0 : -1;
}

static const source_def blitzortung_lightning_def = {
  .id = "blitzortung-lightning", .collector = "environment",
  .name = "Blitzortung lightning network", .name_ja = "Blitzortung 雷検知",
   .update_interval_sec = 60, .run = run };
REGISTER_SOURCE(blitzortung_lightning_def)
